components/script/dom/globalscope.rs

```diff
@@ -2596,59 +2596,55 @@ impl GlobalScope {
-                            if script_private.is_undefined() {
-                                debug!("Set script private for {}", script_base_url);
-
-                                let module_script_data = Rc::new(ModuleScript::new(
-                                    script_base_url,
-                                    fetch_options,
-                                    // We can't initialize an module owner here because
-                                    // the executing context of script might be different
-                                    // from the dynamic import script's executing context.
-                                    None,
-                                ));
-
-                                SetScriptPrivate(
-                                    *compiled_script,
-                                    &PrivateValue(Rc::into_raw(module_script_data) as *const _),
-                                );
-                            }
-
-                            debug!("evaluating Dom string");
-                            JS_ExecuteScript(*cx, compiled_script.handle(), rval)
                         },
-                        SourceCode::Compiled(compiled_script) => {
-                            let script = compiled_script.source_code.get();
-                            let script_handle = HandleScript::new(&script);
-                            JS_ExecuteScript1(*cx, script_handle)
+                        SourceCode::Compiled(pre_compiled_script) => {
+                            compiled_script.set(pre_compiled_script.source_code.get());
                         },
                     };
 
+                    rooted!(in(*cx) let mut script_private = UndefinedValue());
+                    JS_GetScriptPrivate(*compiled_script, script_private.handle_mut());
+
+                    // When `ScriptPrivate` for the compiled script is undefined,
+                    // we need to set it so that it can be used in dynamic import context.
+                    if script_private.is_undefined() {
+                        debug!("Set script private for {}", script_base_url);
+
+                        let module_script_data = Rc::new(ModuleScript::new(
+                            script_base_url,
+                            fetch_options,
+                            // We can't initialize an module owner here because
+                            // the executing context of script might be different
+                            // from the dynamic import script's executing context.
+                            None,
+                        ));
+
+                        SetScriptPrivate(
+                            *compiled_script,
+                            &PrivateValue(Rc::into_raw(module_script_data) as *const _),
+                        );
+                    }
+
+                    let result = JS_ExecuteScript(*cx, compiled_script.handle(), rval);
```

Comment:
Because of the early return in `SourceCode::Text` branch above and also the null check for returning a network error in `off_thread_compilation_callback`, we should always get a non-null `compiled_script`.

I'm wondering, would it be worth adding an assert before executing the script 👀?

Reply:
Sounds reasonable!